### PR TITLE
Fix indexing for overrides.

### LINKED ISSUE #46277: `__torch_function__` does not get call for torch.Tensor's `__getitem__` syntax
## 🐛 Bug

If you write `tensor_val[my_custom_type_with_torch_function]`, the `__torch_function__` call is not invoked on `my_custom_type_with_torch_function` even though it is defined. This works for other operators like `__add__`
`
## To Reproduce

```
import torch
class MyTensor:
    def __torch_function__(self, orig_method, types, args=None, kwargs=None):
        args = args if args else ()
        kwargs = kwargs if kwargs else {}
        print("CALLED ", orig_method)

a = torch.rand(3, 4)
index = MyTensor()

a + index # works: CALLED <method 'add' of 'torch._C._TensorBase' objects>
a[index] # fails: IndexError: only integers, slices (`:`), ellipsis (`...`), None and long or byte Variables are valid indices (got MyTensor)
```

## Expected behavior

`__getitem__` works the same way as `__add__`

## Environment

```
Collecting environment information...
PyTorch version: 1.8.0a0+f7b9046
Is debug build: True
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.4 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.10.2

Python version: 3.7 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: Quadro GP100
GPU 1: Quadro GP100

Nvidia driver version: 418.116.00
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] bert-pytorch==0.0.1a4
[pip3] numpy==1.17.0
[pip3] torch==1.8.0a0
[pip3] torch-struct==0.4
[pip3] torchfile==0.1.0
[pip3] torchtext==0.8.0a0+db31b5d
[pip3] torchvision==0.8.0a0+fc49090
[pip3] torchviz==0.0.1
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               10.1.243             h6bb024c_0  
[conda] mkl                       2020.1                      217  
[conda] mkl-service               2.3.0            py37he904b0f_0  
[conda] mkl_fft                   1.1.0            py37h23d657b_0  
[conda] mkl_random                1.1.1            py37h0573a6f_0  
[conda] numpy                     1.17.0                   pypi_0    pypi
[conda] torch                     1.8.0a0                   dev_0    <develop>
[conda] torchfile                 0.1.0                    pypi_0    pypi
[conda] torchtext                 0.8.0a0+db31b5d           dev_0    <develop>
[conda] torchvision               0.8.0a0+fc49090           dev_0    <develop>
[conda] torchviz                  0.0.1                    pypi_0    pypi
```

## Additional context

This is causing issues while symbolically tracing models with torch.fx, e.g.:
https://github.com/pytorch/benchmark/blob/5768afa082e8931424f803edfca1ac4a627b6143/torchbenchmark/models/Super_SloMo/slomo_model.py#L325

### PR DESCRIPTION
**Pull request by hameerabbasi**
Fixes #46277

### CODE REVIEW

**Comment by hameerabbasi** [comment]
All the `clang-tidy` failures seem to be in `aten/src/ATen/TensorIndexing.h`, which I didn't touch in this PR.

**Comment by ezyang** [comment]
Would be great to get the instruction count delta on this change. @hameerabbasi do you know how to get it?

**Comment by hameerabbasi** [comment]
Unfortunately not.

**Comment by robieta** [comment]
Give this a shot:
```
import torch
from torch.utils.benchmark import Timer


# We have to use a setup string for Callgrind because it uses subprocess.
mytensor_setup = """
class MyTensor:
    def __torch_function__(self, orig_method, types, args=None, kwargs=None):
        args = args if args else ()
        kwargs = kwargs if kwargs else {}

index = MyTensor()
"""

for name, setup in (('Tensor', 'index = torch.LongTensor([1])'), ('MyTensor', mytensor_setup)):
    try:
        c = Timer(
            stmt='a[index]',
            setup=f'{setup}\na = torch.rand(3, 4)'
        ).collect_callgrind(number=10_000, collect_baseline=False)
    except IndexError:
        continue
    print(f"{name:<10}  {c.counts(denoise=True)}")
```

**Comment by hameerabbasi** [comment]
Thanks!

## Benchmarks
### The Script
```
import torch
from torch.utils.benchmark import Timer


# We have to use a setup string for Callgrind because it uses subprocess.
mytensor_setup = """
class MyTensor(torch.Tensor):
    def __torch_function__(self, orig_method, types, args=None, kwargs=None):
        if kwargs is None:
            kwargs = {}
        return super().__torch_function__(orig_method, types, args, kwargs)
"""

for name, setup in (('Tensor', 'a = torch.LongTensor([0])'), ('MyTensor', "a = torch.LongTensor([0]).as_subclass(MyTensor)")):
    for namei, setupi in (('Tensor', 'index = torch.LongTensor(0)'), ('MyTensor', "index = torch.LongTensor(0).as_subclass(MyTensor)")):
        stmt=f'{name}[{namei}]'
        try:
            c = Timer(
                stmt='a[index]',
                setup=f'{mytensor_setup}\n{setup}\n{setupi}'
            ).collect_callgrind(number=10_000, collect_baseline=False)
            print(f"{stmt:<20}  {c.counts(denoise=True)}")
        except Exception:
            print(f"{stmt:<20}  Unsuccessful")
```


### Without this PR
(Note that `Tensor[MyTensor]` had incorrect behavior before, now it's going through `__torch_function__`).
```
Tensor[Tensor]        1035322077
Tensor[MyTensor]      1037292729
MyTensor[Tensor]      1420896375
MyTensor[MyTensor]    1423254494
```

### With this PR
```
Tensor[Tensor]        1033741973
Tensor[MyTensor]      1491081941
MyTensor[Tensor]      1471777249
MyTensor[MyTensor]    1492748973
```

**Comment by ezyang** [comment]
`Tensor[Tensor]` case looks good and that's what I was concerned about. Thanks for testing.

**Comment by hameerabbasi** [comment]
@ezyang I've rebased. The only failure is `clang-tidy`, which I've responded to in https://github.com/pytorch/pytorch/pull/49324#issuecomment-744543901

Is there anything else needed to get this in?

**Comment by ezyang** [comment]
Thanks for bumping, I'd forgotten about this

File: torch/csrc/utils/python_arg_parser.cpp [hunk 1]
```diff
@@ -262,6 +262,28 @@ auto handle_torch_function(PythonArgs &r, PyObject* args, PyObject* kwargs, PyOb
   return handle_torch_function(r, nullptr, args, kwargs, torch_api, module_name);
 }
 
+auto handle_torch_function_indexing(PyObject* self, PyObject* index, PyObject* val) -> PyObject* {
+  std::string func_name = (val == nullptr) ? "__getitem__" : "__setitem__";
```
Comment (ezyang) [inline torch/csrc/utils/python_arg_parser.cpp:266]: You don't need a full std::string here, a char* is enough

**Review by ezyang — APPROVED** [review]
Nice work, thank you!

File: test/test_overrides.py [hunk 2]
```diff
@@ -878,5 +878,72 @@ def f(a):
...
+
+    def test_getitem_subclass(self):
+        class A(torch.Tensor):
+            @classmethod
+            def __torch_function__(cls, func, types, args, kwargs=None):
+                return -1
+
+        t = torch.tensor([5])
+        self.assertEqual(t[A()], -1)
+        self.assertEqual(t[A(), A()], -1)
```
Comment (ezyang) [inline test/test_overrides.py:901]: A more convincing test here would be if the first index was a normal number (to show that we are testing all indices)

Comment (hameerabbasi) [reply test/test_overrides.py]: Done; including all the other locations.

File: torch/csrc/utils/python_arg_parser.cpp (hunk 1 above)
Comment (hameerabbasi) [reply torch/csrc/utils/python_arg_parser.cpp]: Done!

**Comment by hameerabbasi** [comment]
The breaking jobs aren't related to this PR. One is about a backwards-incompatible `aten::sort`, the second about `ModuleNotFoundError: No module named 'cloud_tpu_client'`.

**Pull request closed without merging**